substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jdk/ThreadMXUtils.java

```diff
@@ -0,0 +1,158 @@
+import com.oracle.svm.core.thread.PlatformThreads;
+
+import java.lang.management.ThreadInfo;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.concurrent.locks.AbstractOwnableSynchronizer;
+import java.util.concurrent.locks.LockSupport;
+
+/**
+ * Utils to support {@link SubstrateThreadMXBean} for providing threading information for JMX
+ * support. Include the {@link ThreadInfo} constructing utils, and deadlock detection utils.
+ */
+public class ThreadMXUtils {
+
+    public static class ThreadInfoConstructionUtils {
+
+        private static StackTraceElement[] getStackTrace(Thread thread, int maxDepth) {
+            StackTraceElement[] stackTrace = thread.getStackTrace();
+            return maxDepth == -1 || maxDepth >= stackTrace.length ? stackTrace : Arrays.copyOfRange(stackTrace, 0, maxDepth);
+        }
+
+        private record Blocker(Object blockObject, Thread ownerThread) {
+        }
+
+        private static Blocker getBlockerInfo(Thread thread) {
+            Object blocker = LockSupport.getBlocker(thread);
+
+            if (blocker instanceof JavaMonitor javaMonitor) {
+                return new Blocker(
+                                javaMonitor.getBlockedObject(),
+                                SubstrateThreadMXBean.getThreadById(javaMonitor.getOwnerThreadId()));
+            } else if (blocker instanceof AbstractOwnableSynchronizer synchronizer) {
+                return new Blocker(synchronizer,
+                                SubstrateUtil.cast(synchronizer, Target_java_util_concurrent_locks_AbstractOwnableSynchronizer.class)
+                                                .getExclusiveOwnerThread());
+            }
+            return new Blocker(blocker, null);
+        }
+
+        private static Object[] getLockedSynchronizers(Thread thread) {
+            return JMXMonitoring.getThreadLocks(thread);
+        }
+
+        private record LockedMonitors(Object[] monitorObjects, int[] monitorDepths) {
+        }
+
+        private static LockedMonitors getLockedMonitors(Thread thread, int stacktraceLength) {
+            List<JMXMonitoring.MonitorInfo> monitors = JMXMonitoring.getThreadMonitors(thread);
+            Object[] monitorObjects = monitors.stream().map(JMXMonitoring.MonitorInfo::originalObject).toArray();
+            int[] monitorDepths = monitors.stream().mapToInt(monitorInfo -> stacktraceLength < 0 ? -1 : stacktraceLength - monitorInfo.stacksize()).toArray();
+            return new LockedMonitors(monitorObjects, monitorDepths);
+        }
+
+        private static int getThreadState(Thread thread, boolean inNative) {
+            int state = PlatformThreads.getThreadStatus(thread);
+            if (inNative) {
+                // set the JMM thread state native flag to true:
+                state |= 0x00400000;
+            }
+            return state;
+        }
+
+        public static ThreadInfo getThreadInfo(Thread thread, int maxDepth,
```

Comment:
What should this method return if the thread is no longer
a. alive
b. blocked

I think a lot of corner cases are missing here.